File: test-drive-xml-sax/src/mainjava/com/steelzack/test/drive/books/xml/SAXBooksParser.java
package com.steelzack.test.drive.books.xml;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

public class SAXBooksParser extends DefaultHandler {
	private static final String AUTHOR = "author";

	private static final String BOOK = "book";

	private static final String GENRE = "genre";

	private static final String PAGES = "pages";

	private static final String NAME = "name";

	private String lastValue = null;

	private Book currentBook = null;

	private List<Book> books = new ArrayList<>();

	public List<Book> getBooks() {
		return books;
	}

	@Override
	public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
		super.startElement(uri, localName, qName, attributes);
		if (localName == BOOK) {
			currentBook = new Book();
			currentBook.setTitle(attributes.getValue(NAME));
		}
	}

	@Override
	public void endElement(String uri, String localName, String qName) throws SAXException {
		super.endElement(uri, localName, qName);
		if (localName.equals(PAGES)) {
			currentBook.setPages(Integer.parseInt(lastValue));
		} else if (localName.equals(GENRE)) {
			currentBook.setGenre(lastValue);
		} else if (localName.equals(AUTHOR)) {
			currentBook.setAuthor(lastValue);
		} else if (localName.equals(BOOK)) {
			books.add(currentBook);
		}
	}

	@Override
	public void characters(char[] ch, int start, int length) throws SAXException {
		super.characters(ch, start, length);
		lastValue = new String(Arrays.copyOfRange(ch, start, start + length));
	}
}

File: xml-performance/test-drive-xml-sax/src/main/kotlin/org/jesperancinha/test/drive/sax/books/SAXBooksParser.kt
package org.jesperancinha.test.drive.sax.books

import org.xml.sax.Attributes
import org.xml.sax.SAXException
import org.xml.sax.helpers.DefaultHandler
import java.util.*

class SAXBooksParser : DefaultHandler() {
    private var lastValue: String? = null
    private var currentBook: Book? = null
    private val books: MutableList<Book> = ArrayList()
    fun getBooks(): List<Book> {
        return books
    }

    @Throws(SAXException::class)
    override fun startElement(uri: String, localName: String, qName: String, attributes: Attributes) {
        super.startElement(uri, localName, qName, attributes)
        if (localName === BOOK) {
            currentBook = Book()
            currentBook?.title = attributes.getValue(NAME)
        }
    }

    @Throws(SAXException::class)
    override fun endElement(uri: String, localName: String, qName: String) {
        super.endElement(uri, localName, qName)
        when (localName) {
            PAGES -> currentBook?.pages = lastValue?.toInt() ?: 0
            GENRE -> currentBook?.genre = lastValue
            AUTHOR -> currentBook?.author = lastValue
            BOOK -> currentBook?.let { books.add(it) }
        }
    }

    @Throws(SAXException::class)
    override fun characters(ch: CharArray, start: Int, length: Int) {
        super.characters(ch, start, length)
        lastValue = String(ch.copyOfRange(start, start + length))
    }

    companion object {
        private const val AUTHOR = "author"
        private const val BOOK = "book"
        private const val GENRE = "genre"
        private const val PAGES = "pages"
        private const val NAME = "name"
    }
}